migrate.go

```diff
@@ -887,13 +887,44 @@ func (m *Migrate) unlock() error {
 	m.isLockedMu.Lock()
 	defer m.isLockedMu.Unlock()
 
-	if err := m.databaseDrv.Unlock(); err != nil {
-		// BUG: Can potentially create a deadlock. Add a timeout.
-		return err
-	}
+	// create done channel, used in the timeout goroutine
+	done := make(chan bool, 1)
+	defer func() {
+		done <- true
+	}()
 
-	m.isLocked = false
-	return nil
+	// use errchan to signal error back to this context
+	errchan := make(chan error, 2)
+
+	// start timeout goroutine
+	timeout := time.After(m.LockTimeout)
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			case <-timeout:
+				errchan <- ErrLockTimeout
+				return
+			}
+		}
+	}()
+
+	// now try to unlock
+	go func() {
+		if err := m.databaseDrv.Unlock(); err != nil {
```

Comment:
The race detector went off and I believe calling `m.databaseDrv.Unlock()` in a goroutine is the culprit since the the db driver is now being accessed unlocked e.g. outside `m.isLockedMu`

Reply:
The mutex should stay locked until the goroutines are finished running so I dont think thats the case.  The lock method works the same way.  What I think I'll do to make things easier is to split this MR into two things.  So we can discuss each item seperately.